### PR TITLE
feat: introduce `first_opt()`, `first_key_opt()`, `last_opt()`, `last_key_opt()` and `search_opt()`

### PR DESCRIPTION
**Pull request by simPod**
I'd like to modify search operations over `iterables` by wrapping the result in Option.

Currently, it is not possible to use these functions to find `null` value since `null` represents "no result".

This applies to all `first*`, `last*` and `search` functions.

WDYT?

### CODE REVIEW

**Comment by azjezz** [comment]
I like the idea of returning an `Option`, but i dislike having this type of BC, what about adding new functions that result in an `Option`? e.g `first_opt<T>(iterable<T> $i): Option<T>` vs `first<T>(iterable<T> $I): null|T`?

`_opt` suffix is one way to go about this ( this is a popular choice in the rust community, e.g: https://docs.rs/chrono/latest/chrono/struct.Date.html#method.and_hms_micro VS https://docs.rs/chrono/latest/chrono/struct.Date.html#method.and_hms_micro_opt), but i'm open to suggestions if you have any, maybe we can find a more suitable naming convention for those functions.

cc @veewee this might interest you.

**Comment by simPod** [comment]
I'll go for it. Thanks for input.

**Comment by veewee** [comment]
@azjezz 

The migration path might be better this way indeed.
However you'll probably end up to something that looks like this in the long run, which might be something we'dd want to avoid?

* v3 -> deprecate old API in favour for `_opt`
* v4 -> remove old API
* v5 -> Recreate original function without the `_opt` suffix and deprecate the `_opt` suffix functions
* v6 -> remove the `_opt` functions again

**Comment by azjezz** [comment]
@veewee personally i would like to keep both, i don't see why we need to deprecate/remove the old behaviour.

File: src/Psl/Iter/last_key_opt.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * if the iterable is empty, {@see Option::none} will be returned.
+ *
+ * @template Tk
+ * @template Tv
+ *
+ * @param iterable<Tk, Tv> $iterable
+ *
+ * @return Option<Tk>
+ */
+function last_key_opt(iterable $iterable)
```
Comment (azjezz) [inline src/Psl/Iter/last_key_opt.php]:
```suggestion
function last_key_opt(iterable $iterable): Option
```

File: src/Psl/Iter/last_key_opt.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    $last = Option::none();
+    foreach ($iterable as $k => $_) {
+        $last = $k;
+    }
+
+    return $last instanceof Option ? $last : Option::some($last);
```
Comment (azjezz) [inline src/Psl/Iter/last_key_opt.php]: this breaks static analysis.

if the given type is `iterable<string, Option<bool>>` psalm assume the return type is `Option<Option<bool>>`, you are breaking that here by making it `Option<bool>`. meaning, regardless of the type of `$last` it should be returned as is.

```suggestion
    $last = Option::none();
    foreach ($iterable as $k => $_) {
        $last = Option::some($k);
    }

    return $last;
```

File: src/Psl/Iter/last_opt.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * @return Option<Tv>
+ */
+function last_opt(iterable $iterable)
+{
+    $last = Option::none();
+    foreach ($iterable as $v) {
+        $last = $v;
+    }
+
+    return $last instanceof Option ? $last : Option::some($last);
```
Comment (azjezz) [inline src/Psl/Iter/last_opt.php]: same as above.

**Review by azjezz — CHANGES_REQUESTED** [review]

File: src/Psl/Iter/last_key_opt.php (hunk 2 above)
Comment (simPod) [reply src/Psl/Iter/last_key_opt.php]: true. i did not want to initialize Option for each value but you're right

File: src/Psl/Iter/last_opt.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ * Returns the last element of an iterable wrapped in {@see Option::some},
+ * if the iterable is empty, {@see Option::none} will be returned.
+ *
+ * @template Tv
+ *
+ * @param iterable<Tv> $iterable
+ *
+ * @return Option<Tv>
+ */
+function last_opt(iterable $iterable)
```
Comment (azjezz) [inline src/Psl/Iter/last_opt.php]: missing return type.
```suggestion
function last_opt(iterable $iterable): Option
```

File: src/Psl/Iter/search_opt.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ *      => Null
+ *
+ * @template T
+ *
+ * @param iterable<T> $iterable The iterable to search
+ * @param (Closure(T): bool) $predicate
+ *
+ * @return Option<T>
+ */
+function search_opt(iterable $iterable, Closure $predicate)
```
Comment (azjezz) [inline src/Psl/Iter/search_opt.php]: same

**Review by azjezz — CHANGES_REQUESTED** [review]

File: src/Psl/Iter/search_opt.php (hunk 5 above)
Comment (simPod) [reply src/Psl/Iter/search_opt.php]:
```suggestion
function search_opt(iterable $iterable, Closure $predicate): Option
```

File: src/Psl/Iter/search_opt.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * the value of the matching element wrapped in {@see Option::some}.
+ * If a predicate never returns true, {@see Option::none} will be returned.
+ *
+ * Examples:
+ *
+ *      Iter\search(['foo', 'bar', 'baz'], fn($v) => 'baz' === $v)
+ *      => Str('baz')
+ *
+ *      Iter\search(['foo', 'bar', 'baz'], fn($v) => 'qux' === $v)
+ *      => Null
```
Comment (azjezz) [inline src/Psl/Iter/search_opt.php]: exampless are outdated

**Review by azjezz — CHANGES_REQUESTED** [review]
one last thing 😅

File: src/Psl/Iter/search_opt.php (hunk 6 above)
Comment (simPod) [reply src/Psl/Iter/search_opt.php]: good catch. is it correct now? Not sure what `Str('baz')` was, replaced.

**Review by azjezz — APPROVED** [review]

**Review by veewee — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**